temporian/core/operators/until_next.py

```diff
@@ -0,0 +1,113 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""UntilNext operator class and public API function definitions."""
+
+import math
+
+from temporian.core import operator_lib
+from temporian.core.compilation import compile
+from temporian.core.data.node import (
+    EventSetNode,
+    create_node_new_features_new_sampling,
+)
+from temporian.core.operators.base import Operator
+from temporian.core.typing import EventSetOrNode
+from temporian.proto import core_pb2 as pb
+from temporian.utils.typecheck import typecheck
+from temporian.core.data.dtype import DType
+from temporian.core.data.duration_utils import (
+    Duration,
+    NormalizedDuration,
+    normalize_duration,
+)
+
+
+class UntilNext(Operator):
+    def __init__(
+        self,
+        input: EventSetNode,
+        sampling: EventSetNode,
+        timeout: NormalizedDuration,
+    ):
+        super().__init__()
+
+        self.add_input("input", input)
+        self.add_input("sampling", sampling)
+
+        self.add_attribute("timeout", timeout)
+        self._timeout = timeout
+
+        if not math.isfinite(timeout):
+            raise ValueError(
+                f"Timeout should be finite. Instead, got {timeout}"
+            )
+
+        if timeout <= 0:
+            raise ValueError(
+                f"Timeout should be strictly positive. Instead, got {timeout}"
+            )
+
+        self.add_output(
+            "output",
+            create_node_new_features_new_sampling(
+                features=[("until_next", DType.FLOAT64)],
+                indexes=input.schema.indexes,
+                is_unix_timestamp=input.schema.is_unix_timestamp,
+                creator=self,
+            ),
+        )
+
+        self.check()
+
+    @property
+    def timeout(self) -> float:
+        return self._timeout
+
+    @classmethod
+    def build_op_definition(cls) -> pb.OperatorDef:
+        return pb.OperatorDef(
+            key="UNTIL_NEXT",
+            attributes=[
+                pb.OperatorDef.Attribute(
+                    key="timeout",
+                    type=pb.OperatorDef.Attribute.Type.FLOAT_64,
+                ),
+            ],
+            inputs=[
+                pb.OperatorDef.Input(key="input"),
+                pb.OperatorDef.Input(key="sampling"),
+            ],
+            outputs=[pb.OperatorDef.Output(key="output")],
+        )
+
+
+operator_lib.register_operator(UntilNext)
+
+
+@typecheck
+@compile
+def until_next(
+    input: EventSetOrNode,
+    sampling: EventSetNode,
+    timeout: Duration,
+) -> EventSetOrNode:
+    assert isinstance(input, EventSetNode)
+
+    return UntilNext(
+        input=input,
+        sampling=sampling,
+        timeout=normalize_duration(timeout),
```

Comment:
is there any reason to call `normalize_duration` here instead of in the class's constructor? I moved it to the constructor in the window ops

Reply:
The "timeout" argument is tagged as FLOAT_64 in the op.
It means the constructor should receive a float i.e., the normalization should be done before.

Reply:
Yes.

I think that informally, my concern is that checking should be done after normalization.
Imagine the following bad case:

- User feed some input
- Input is checked and validated.
- Input is normalized to something else.
- Normalized input is serialized.
- Later, normalized input is deserialized.
- Normalized input is checked and **invalid**
- The user is blocked.

Instead, it is better to have:

- User feed some input
- Input is normalized to something else.
- Normalized input is checked and **invalid**

This is especially problematic because while normalized input are very constrained, user inputs can be a lot of things.

Having the operator class being super restrictive (i.e., only allowing normalized inputs) and doing the checks is a way to ensure we are always in the second case.

If you are not convinced, let me know. I am happy to update this PR.

Reply:
Okay, got the idea, but I'd argue that having all the op's logic centralized in the class would be desirable. Although if we very clearly define what goes in the function and what goes in the class that'd be fine by me too.

> Normalized input is checked and invalid

When/why would this happen? I think this is the main issue - shouldn't normalized inputs always pass any check we're doing on them? On second thought, I guess this could be a problem if the normalization we're doing isn't idempotent (since we'd be applying it twice, once when op is instantiated and again when its deserialized), but it should be.

Another option would be to have the op's main constructor do normalization, and have a secondary one that is used by the deserialization code that skips the normalization and checks, assuming they will pass.

Reply:
>> On second thought, I guess this could be a problem if the normalization we're doing isn't idempotent 

That is the issue :). We have really poor test coverage for that (I suspect we actually never test it).

>> Another option would be to have the op's main constructor do normalization, and have a secondary one that is used by the deserialization code that skips the normalization and checks, assuming they will pass.

There can only be one constructor in python.

Another alternative is to unit test operator stability to serialization. For instance, we could unit test that an operator attributes are the same with and without being serialized+deserialized.

Reply:
Feel free to change the window ops to do the normalization in the function call too to keep it consistent